### ISSUE TITLE
Ansible 2.2 can only access first host's facts when adhoc group contain many hosts

### ISSUE THREAD

**Issue by stanrate** [ansible/ansible#19274]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

hostvars

##### ANSIBLE VERSION

```
ansible 2.2.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### STEPS TO REPRODUCE

```
- name: Populate host group with host names
# hosts: host1:host2:host3:hosts4
  hosts: host1:host2:host3:hosts4:host5:host6:host7:host8
  gather_facts: yes
  tasks:
  - name: task1
    group_by: key='adhoc_group'


- name: Run playbook on all hosts in host group "adhoc_group"
  hosts: adhoc_group
  gather_facts: no
  tasks:
  - name: task2
    debug: var=inventory_hostname


- name: Show first host's architecture
  hosts: localhost
  gather_facts: no
  tasks:
  - name: task3
    debug: var=hostvars[groups['adhoc_group'][0]]['ansible_architecture']


- name: Show second host's architecture
  hosts: localhost
  gather_facts: no
  tasks:
  - name: task4
    debug: var=hostvars[groups['adhoc_group'][1]]['ansible_architecture']
```

##### ACTUAL RESULTS

```
TASK [task4] *******************************************************************
ok: [localhost] => {
    "hostvars[groups['adhoc_group'][1]]['ansible_architecture']": "VARIABLE IS NOT DEFINED!"
}
```

**Comment by stanrate** [comment]
When the playbook is run on Ansible 2.2 with four hosts, the following correct output is produced:
TASK [task4] *******************************************************************
ok: [localhost] => {
    "hostvars[groups['adhoc_group'][1]]['ansible_architecture']": "x86_64"
}
When the playbook is run on Ansible 1.9.4 or 2.1, the above correct output is also produced with a large number of hosts.
When the playbook is run on Ansible 2.2 with eight or more hosts, message "VARIABLE IS NOT DEFINED!" is issued.

**Comment by vazhnov** [comment]
- Xubuntu 16.04 amd64
- Python 2.7.11-1
- Ansible 2.2.0.0-1ppa~xenial
- 584 hosts in "{{ groups['all']|length }}"

Module group_by is not works for me, but works fine in 1.9.6-1ppa~vivid.

Code:

```
- name: 'Create dynamic groups by cluster name'
  group_by:
    key: "dyn_group_cluster_{{ cluster_name }}"
```

Verbose output:

```
TASK [aws_common : Create dynamic groups by cluster name] ***
task path: /home/user1/git/ansible/devops-ansible/roles/aws_common/tasks/set_facts.yml:28
ok: [idev-04-data03.test.example.com] => {
    "add_group": "dyn_group_cluster_idev-04-data", 
    "changed": false, 
    "invocation": {
        "module_args": {
            "key": "dyn_group_cluster_idev-04-data"
        }, 
        "module_name": "group_by"
    }
}
```

But "{{ groups['dyn_group_cluster_idev-04-data'] }}" is not defined.

**Comment by stanrate** [comment]
Since the group_by module seems to no longer work properly in Ansible 2.2, I tried to achieve the same result by using the add_host module and it worked. I replaced the following:

group_by: key='adhoc_group'

with:

 add_host: hostname={{item}} groups='adhoc_group'
 with_items: "{{ansible_play_hosts}}"

and it worked. 

It is fortunate that there are two ways of creating dynamic adhoc groups of hosts, one using "group_by" module which no longer seems to work properly in Ansible 2.2 and the "add_hosts" module that still works.

**Comment by bcoca** [comment]
group_by 'reloads' host vars, add_host does not, it just appends, this is probably the difference you are seeing here.

leaving ticket open to consider if the 'reload' should/should'nt include facts.

**Comment by ghost** [comment]
**IGNORE**:
This problem is the now "classic" misstep, using an inventory file in a directory that does NOT have host_vars and groups_vars as a subdirectory.

**Known problem description, for those who don't understand this yet, for this run I have an inventory at /etc/ansible/hosts and host_vars/group_vars in current directory:**
This may be a similar problem, hostvars is changing based on number of hosts.

version 2.2.1.0 , Centos package

 ---
```
- hosts: one
  roles
    - debug
  tasks:
    - debug: msg="two {{hostvars['two']['boolean']}}"
    - debug: msg="three {{hostvars['three']['boolean']}}"
    - debug: msg="four {{hostvars['four']['boolean']}}"

```
Role debug has the same tasks, this is to test if role makes any difference.

With boolean false for two and true for three and four, with a single host in hosts, output is correct, false is shown for two. With change `- hosts: all` and run for single host (`-l one`), where my all group contains 114 hosts, then true is shown for two in both messages.

Is this a known problem that I've missed finding? Related to this problem?

UPDATE:

Any --limit, including playbook with single host, shows the change in host_vars.

In addition, to clarify why no "undefined" is noted, a default as false is defined via groups_vars.

**Comment by stanrate** [comment]
In Ansible 2.2.1, the group_by bug seems to be fixed because I can't reproduce it. Ansible 2.2.0 was the only release where group_by seemed to work but if a task tried to use a fact belonging to a host that had been added to an adhoc group (using group_by), a "VARIABLE IS NOT DEFINED!" message was emitted. I ran the same playbook that failed in 2.2.0 on 2.2.1 and cannot reproduce the problem. 

I read bcoca 6 January comment that states "group_by 'reloads' host vars, add_host does not, it just appends" and ran ansible-doc for group_by and for add_host but there was no mention of reloading or appending host_vars. As far as I'm concerned, if a host belongs to multiple groups in the inventory and host vars exist for that host and a task uses group-by to create an additional group, there is no reason why the host's host vars should be affected in any way.
